ENH: Added atleast_nd

madphysicist opened the pull request:
Introduced generalized `atleast_nd` function. Similar to `atleast_1d`, `atleast_2d`, `atleast_3d`. Also added for masked arrays. Tests included.

This PR also fixes an issue with masked arrays that would crash the `atleast_*` functions if multiple scalars were input, e.g. `np.ma.atleast_1d(1.0, 2.0, 3.0)` would raise an error.

madphysicist committed:
ENH: Added atleast_nd

Function only accepts one array and an indexed placement spec.
Masked arrays fully supported.
atleast_1d and atleast_2d re-implemented in terms of new func.

Tests included.

mhvk commented on numpy/core/shape_base.py:
```diff
@@ -177,6 +177,71 @@ def atleast_3d(*arys):
+    """
+    res = []
+    for ary in arys:
+        ary = asanyarray(ary)
```
As noted on the mailing list, if we go for this at all, you can just write
```
ary = array(ary, copy=False, subok=True, ndim=n)
```

madphysicist replied:
Not if I add an option to place the dimensions at the end. I may still work this in without the `ndim` argument for that case though.

madphysicist replied:
I did not use your suggestion literally because zero-dim arrays need special handling, but I did something quite similar, just without the `ndmin` parameter (it's not `ndim` for `array`).

mhvk commented on numpy/core/shape_base.py:
```diff
@@ -177,6 +177,71 @@ def atleast_3d(*arys):
+
+    Parameters
+    ----------
+    n : scalar
```
I think the large number of changes in `np.ma` suggest this order is not the right one. I would suggest to instead make `n` a keyword-only argument, and flip the order.

madphysicist replied:
I disagree. I think it is important to have `n` as the first parameter to `atleast_nd` from an intuitive point of view. That parameter is required and does not make nearly as much sense at the end of the list.

The changes in `ma` are not very drastic and don't affect the existing cases. I basically added the option to have a "prefix" of non-array args to a class that is private to begin with. All the existing tests should run fine now that I have fixed the syntax for pre-3.5 Python.

mhvk replied:
But almost all numpy functions have the array on which the function works first in the list. The only real reason for doing it differently here is that one is allowed to pass in multiple arrays as `*args` (a design mistake, in my opinion, but too late...).

Anyway, these are just my opinions; as I'm not sure I understand the need for the function, I probably should not be the one arguing and rather leave it to the numpy maintainers to decide.

madphysicist replied:
I don't disagree with the possibility that this is a design mistake. The function should only take one array argument. If you need to pass in a tuple or list, use an iterator or list comprehension or whatever. In fact, let me think about it and perhaps I will break away from the existing convention for the atleast functions.

The need for the function comes from a use-case that I ran across. Perhaps it is not that universal, but I think it will be useful nevertheless.

madphysicist replied:
I am reworking the tests and came across a case that shows that the `ma` stuff *has* to be reworked: Making `b = atleast_1d(1.0)` into `b = atleast_1d(1.0, 2.0)` in `TestShapeBase.test_shape_scalar` (first test) would actually crash it since it would enter the wrong conditional in `_fromnxfunction`. Update to that coming soon.

madphysicist commented:
So yeah...

madphysicist commented:
bump

juliantaylor commented:
can you implement the lower dimension variants in terms of this one?

madphysicist commented:
Not `atleast_3d` until I introduce a more complex spec for adding dimensions. `atleast_3d` adds dimensions on both sides of a 1D array. The other two are fine.

juliantaylor commented:
unless this function is significantly slower I would suggest to implement 1d and 2d in terms of this one for the extra test coverage and less code.

madphysicist commented:
Done. Reduced `atleast_[12]d` to three lines of code. Could have probably been one, but the input parameters are weird.

homu commented:
:umbrella: The latest upstream changes (presumably #7823) made this pull request unmergeable. Please resolve the merge conflicts.

shoyer commented on numpy/core/shape_base.py:
```diff
@@ -177,6 +160,82 @@ def atleast_3d(*arys):
+    -------
+    res : ndarray
+        An array with ``a.ndim >= n``. Copies are avoided where
+        possible, and a view with `n` or more dimensions is returned.
```
Is a copy ever needed?

madphysicist replied:
I do not know enough about the internals to know if `res = array(ary, copy=False, subok=True)` will ever return a copy under normal circumstances. My guess is no, but I wanted to do some mild CYA in the docs, just in case.

shoyer replied:
It will only make a copy for non numpy arrays (e.g., lists).
On Mon, Aug 29, 2016 at 8:12 AM Joseph Fox-Rabinovitz <
notifications@github.com> wrote:

> In numpy/core/shape_base.py
> <https://github.com/numpy/numpy/pull/7804#discussion_r76622781>:
>
> > +    ----------
> > +    ary : array_like
> > +        The input array. Non-array inputs are converted to arrays.
> > +        Arrays that already have `n` or more dimensions are preserved.
> > +    n : scalar
> > +        The minimum number of dimensions required for each of the input
> > +        arrays.
> > +    pos : {'start', 'end'}, optional
> > +        Where to add new dimensions. All new dimensions are added on the
> > +        same side of the shape. Default is 'start'.
> > +
> > +    Returns
> > +    -------
> > +    res : ndarray
> > +        An array with ``a.ndim >= n``. Copies are avoided where
> > +        possible, and a view with `n` or more dimensions is returned.
>
> I do not know enough about the internals to know if res = array(ary,
> copy=False, subok=True) will ever return a copy under normal
> circumstances. My guess is no, but I wanted to do some mild CYA in the
> docs, just in case.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/7804/files/5b443132ec6c30aef974c73b30b510e6b60e8755#r76622781>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1p0XcxxYEGc_49tnk41VdMQUHXgoks5qkvSFgaJpZM4JFvBv>
> .
>

madphysicist replied:
I have updated the docstring to reflect that now.

homu commented:
:umbrella: The latest upstream changes (presumably #7922) made this pull request unmergeable. Please resolve the merge conflicts.

mattharrigan commented:
see [issue 8214](https://github.com/numpy/numpy/issues/8214)

mattharrigan commented on numpy/core/shape_base.py:
```diff
@@ -177,6 +160,81 @@ def atleast_3d(*arys):
+        Arrays that already have `n` or more dimensions are preserved.
+    n : scalar
+        The minimum number of dimensions required for each of the input
+        arrays.
```
there's only ever 1 array, for each of the input arrays doesn't seem to make sense

madphysicist replied:
Good call. Fixed now.

shoyer commented:
This looks pretty solid to me.  Did we ever discuss it on the mailing list?

Two suggestions:
1. Note that the default `pos='start'` creates arrays that are consistent with NumPy's broadcasting rules.
2. `atleast_3d` should have a loud note that its implementation is inconsistent with the others, for legacy reasons.

madphysicist commented:
Mailing list thread start: https://mail.scipy.org/pipermail/numpy-discussion/2016-July/075759.html. There was no solid consensus to merge, but no strong objections either if I remember correctly.

madphysicist commented:
I think that suggestion 2 is adequately covered in the notes section already. I added another entry for suggestion 1.

shoyer commented:
@madphysicist which "notes" section are you referring to for `atleast_3d`? I see it documented under the "Returns" section but in my opinion that isn't loud enough. I would add it immediately under the docstring, at the top of the function.

madphysicist commented:
My mistake. I added a note to `atleast_nd` because I did not realize you were referring to `atleast_3d`. Fixed by adding a notes section to `atleast_3d`.

shoyer commented:
I see that I even commented on the mailing list discussion!

I guess I've come around and am now slightly in favor of this. Can you post on the mailing list once more so we can come to a final decision? It seemed that reactions were mostly divided between "Why do we have `atleast_?d` in the first place?" and "This would be useful for me".

madphysicist commented:
I will post as a reply to the original thread if you don't have any objections.

homu commented:
:umbrella: The latest upstream changes (presumably #8216) made this pull request unmergeable. Please resolve the merge conflicts.

madphysicist commented:
Rebased onto master, including #8216, which does not affect the atleast_nd function itself.

mattharrigan commented:
Does atleast_3d hint at a useful or common operation?  If so it might be worthwhile to add a third option pos='split', which adds the first newaxis to the right, the second newaxis to the left, the third to the right, and so on.  This would also allow atleast_3d to be an alias for atleast_nd(ary, 3, 'split')

charris commented:
Whoa, that alternating behavior is completely weird, but has been there since the beginning, lo, so many years ago. Does anyone actually use that functionality?

shoyer commented:
My guess is that it exists for the sake of some domain specific conventions around 3D arrays, e.g., for [handling images](http://scikit-image.org/docs/dev/user_guide/numpy_images.html#coordinate-conventions).

`pos='split'` sounds crazy to me. There's no generally useful pattern here, just domain specific conventions. If an imaging library (or whatever) needs this, they should write their own helper functions.

mattharrigan commented:
I agree it does sound pretty crazy.  I wasn't sure if there were obvious use cases or not.  Sorry for the bad idea.

charris commented:
@shoyer That was my guess also, IIRC David Cournapeau was also working on multidimensional 'cube' iterators in that time period.

madphysicist commented:
We can go arbitrarily crazy with how dimensions are added. For example, an iterable of ints that specify the index at which to add the next dimension, or a dictionary that does the same, or a dict of iterables, keyed by the starting number of dimensions. I don't think that implementing any of these has much use. `pos='start'` and `pos='end'` ought to cover 99.9999% of common use cases as it is.

charris commented:
The general usefulness of these functions is that they allow the operation to apply to many arrays at the same time. With list comprehension it is pretty easy to extend the dimensions on the left with ones, `[array(x, ndmin=<min dimension>, copy=False) for x in <iterable>]`. What is sometimes useful is extension on the right for broadcasting along rows instead of columns in 2-d element-wise multiplication. Maybe we could add the `ndmin` keyword to `asarray` to make for easier list comprehension.

madphysicist commented:
I can also rewrite this function as `atleast_nd(n, *arys, pos='start')`. This require a bit more code in the `ma` version though.

shoyer commented:
Putting `n` first seems quite strange. However, you could possibly make `n` a required keyword only argument, e.g., `atleast_nd(*arys, n=None, pos='start')`.

madphysicist commented:
I don't think that making `n` a keyword-only argument is a good idea. I also don't think that it's strange to put it first in a function named <tt>atleast_<em><b>n</b></em>d</tt>.

charris commented:
keyword only sounds more natural to me, putting `n` first seems a bit weird. The expectation is that the list of arrays would come first.

charris commented:
I like "left" and "right" better than "start" and "end", the latter is a bit ambiguous as to which end is which.

shoyer commented:
`atleast_nd(x, 2)` is ambiguous at first glance -- it isn't clear whether `2` is an array or the number of dimensions. In contrast, `atleast_nd(x, n=2)` is very clear and only adds two character. So I actually like the required keyword argument. The only downside is that doing this in a Python 2 compatible way breaks argument name introspection.

madphysicist commented:
The reason I like `atleast_nd(2, x)` is that it completely removes that ambiguity while avoiding any potential Python2 issues. I think it makes sense to specify the minimum dimensionality before the array.

madphysicist commented:
Also, when you write `atleast_2d(x)`, the `2` comes before the `x`.

homu commented:
:umbrella: The latest upstream changes (presumably #8182) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented on numpy/core/shape_base.py:
```diff
@@ -177,6 +169,87 @@ def atleast_3d(*arys):
+        raise ValueError('Illegal value of pos: %s' % pos)
+
+    if addons > 0:
+        if pos == 'start' or res.ndim == 0:
```
Is `res.ndim == 0` needed here? I'd prefer it without, so that `np.atleast_nd(a, 0)` always promotes to 0d arrays. Similarly, the line below would become `res[(newaxis,) * addons + (Ellipsis,)]`

eric-wieser commented on doc/release/1.12.0-notes.rst:
```diff
@@ -261,6 +261,11 @@ values. There are two different propagation modes. The default causes masked
 masks if there is no alternative.
 
+Array dimension functions generalized with new ``atleast_nd``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
This should be moved to 1.13.0-notes and changed to `-------` next time it is rebased

eric-wieser commented:
> Maybe we could add the ndmin keyword to asarray to make for easier list comprehension.

Addressed in #10025

madphysicist commented:
@eric-wieser. I will close this down as soon as your PR is accepted.

Erotemic commented:
@madphysicist are you still interested in working on this? I didn't see your PR and I seem to have duplicated some of your efforts in #10879.

Erotemic commented:
I'd like to offer some feedback on the proposed signature:
* Does `atleast_nd` really *have* to accept varargs? I feel like the varargs design in the other functions is clunky, rarely used, and adds unnecessary complexity. 
* I think its much more natural to have `n` as the second argument, but I'm less adamant about this. 
* I agree a boolean is nicer than a string argument. However, I don't think the keyword `append` is intuitive. In my inhouse implementation of this I've used `tofront`, but I'm not convinced that's the best either. I changed it to simply `front` in #10879, but again I'm not convinced its the best. I do like `front` over `append` though. Other alternatives I've considered are: `back`, `left`, `right`, `before`, `after`. `first`, `last`, `toback`, etc... Obviously, I out of these I chose `front` / `tofront`, but I'm wondering if there isn't something better.

eric-wieser commented on numpy/core/shape_base.py:
```diff
@@ -179,6 +170,87 @@ def atleast_3d(*arys):
+
+    if addons > 0:
+        if pos == 'start' or res.ndim == 0:
+            res = res[(newaxis,) * addons]
```
Better to put the `Ellipsis` here too, but on the right

shoyer commented:
I am -1 on adding functions that accept `*args` and return either tuple based upon whether or not `len(args) > 1`. Such functions are quite painful to use in a programmatic way.

madphysicist commented:
The only good argument in favor of varargs is compatibility with the old functions, so I will get rid of it. I still think that it's nice to be able to write `x, y = atleast_nd(3, x, y)`, but I admit that it's not worth complicating the interface for it.

I have two arguments in favor of putting `n` first, besides supporting varargs:

1. I think it makes more sense aesthetically.
2. It makes it much easier to construct a partial function for a particular dimension you are interested in. It is much more likely that you will want to do something like `atleast_100d = partial(atleast_nd, 100)` than applying different dimensions to the same array.

Again, this is not something I'm totally set on, so if the consensus is in favor of putting `n` second, I'll do that.

As far as the side argument goes, I chose `append` because it's clearly assymetrical. It's difficult to choose between pairs like `left/right`, `front/back`, `before/after`, but I feel like `append` is clearly better than `prepend`. I chose that instead of debating whether to go with the default true or default false option from the other candidates. As with everything else, I am perfectly fine following the consensus on this one.

madphysicist commented:
Looking back at this, the reason for having a string as the last argument was to allow more complicated placement schemes than just all front or all back in the future. While I'm not sure it's really worth doing, it's at least worth considering as an option. Comparing strings won't add much overhead, but it will make it much easier to name the argument.

Erotemic commented:
I do prefer a bool over a string, but a string might be the right option simply because it will make the argument more intuitive. I think there should only be 2 modes, though. The third style mentioned in this thread seems completely unnecessary.

madphysicist commented:
I've made the signature `atleast_nd(ary, n, pos='left')`. This should more or less satisfy just about everyone who commented, and looks OK to me. I've changed the `'left'` option to just use `array(..., ndmin=n)` directly. I don't see an immediate problem with doing that, but please let me know if there is one.

Would there be any use to have an `inplace` boolean parameter that would modify the shape of the original object instead of making a view if set to `True`?

eric-wieser commented on numpy/core/shape_base.py:
```diff
@@ -179,6 +170,92 @@ def atleast_3d(*arys):
+    [[[[1, 2, 3]]]]
+    """
+    if pos == 'left':
+        res = array(ary, copy=False, subok=True, ndmin=n)
```
This fails on duck arrays - I'd be inclined to just support anything with a `.ndim` property

eric-wieser replied:
(and go back to the slicing for either case)

madphysicist replied:
But we call `array(ary, copy=False, subok=True)` in every case anyway. I'm not sure I see what the difference would be. I'm fine with going back to slicing for both cases, especially with `pos` about to be an integer now.

eric-wieser commented on numpy/core/tests/test_shape_base.py:
```diff
@@ -125,6 +125,52 @@ def test_3D_array(self):
+        desired_shape = (1, 1, 1, 4, 3)
+        desired_base = a.base  # a was reshaped
+        self.assertEqual(res.shape, desired_shape)
+        self.assertIs(res.base, desired_base)
```
These methods don't exist

eric-wieser commented on numpy/ma/tests/test_extras.py:
```diff
@@ -1565,6 +1565,21 @@ def test_atleast_2d(self):
+        # Test atleast_nd
+        a = masked_array([0, 1, 2], mask=[0, 1, 0])
+        b = atleast_nd(a, 4)
+        c = atleast_nd(a, 4, pos='end')
```
Should be `right`?

madphysicist replied:
Yes. I'm fixing the tests now. Pretty much didn't look at the tests besides borking the imports.

eric-wieser commented:
Not sure it's better, but another option: could do `pos=0` and `pos=-1` instead of left and right, which would look like:

```
if ar.ndim < ndim:
    pos = normalize_axis_index(pos, ar.ndim)
    idx = pos * np.s[:,] + (np.newaxis,)*excess + (arr.ndim - pos) * np.s[:,]
    return arr[idx]
```

Erotemic commented:
I actually like the `pos=0` and `pos=-1` idea a lot. I'm +1 on that.

madphysicist commented:
@eric-wieser I very much like the idea of `pos` being an integer. What do you feel about adding an `inplace` parameter?

madphysicist commented on numpy/ma/tests/test_regression.py:
```diff
@@ -47,7 +47,7 @@ def test_atleast_2d(self):
         # Ticket #1559
         a = np.ma.masked_array([0.0, 1.2, 3.5], mask=[False, True, False])
         b = np.atleast_2d(a)
-        assert_(a.mask.ndim == 1)
```
This change indicates that something is not 100% backwards compatible in the updates I made to `ma`. I believe that the version I have here is the expected and correct one.

eric-wieser replied:
The fact this used to be `1` scares me

madphysicist replied:
I found a number of interesting things in the ma wrapper classes when I first started doing this.

madphysicist replied:
I just realized I made a mistake here: `a.mask.ndim` absolutely should be 1`. And `b.mask.ndim` should be 2. Fixed on my end now. Whoops!

madphysicist commented:
Normalizing the axis actually won't work out of the box. Position `-1` needs to insert *past* the end, not *before* the end. Will have to fiddle with it...

madphysicist commented:
Looks like the PR is in good shape now. The trick was of course to normalize `pos` against `ary.ndim + 1` instead of just `ary.ndim`. The tests and `np.ma` versions look complete as well.

Is there a something I can do about the CircleCI failure? It seems to come up intermittently regardless of anything else, and the message does not appear to be related to anything I did here...

madphysicist commented:
Squashed and ready to go.

eric-wieser commented:
Thinking again, I think we have the wrong meaning of `pos` here. If `arr.shape == (2, 3, 4)`, I think the useful thing to do is allow the output to be `(2, 3, 4, 1, 1)`, `(1, 2, 3, 4, 1)`, or `(1, 1, 2, 3, 4)`.

What you have right now allows `(2, 1, 1, 3, 4)` and `(2, 3, 1, 1, 4)` instead of `(1, 2, 3, 4, 1)`, which I think is less useful

Erotemic commented:
@eric-wieser In your example, I can't think of a situation where `(1, 2, 3, 4, 1)` is a useful thing to do. 

I was thinking that `pos` would simply be restricted to either `0` or `-1`. I think any other case is unlikely to be needed, so it shouldn't be a problem to enforce this restriction and raise a ValueError if pos takes any other value.  

Going back to the bool idea, what about the keyword `end`?

madphysicist commented:
@eric-wieser. The split convention was discussed and thankfully discarded (around [this comment](https://github.com/numpy/numpy/pull/7804/#issuecomment-256974899)). The consensus was that `atleast_3d` is an anomalous case that should be allowed to exist for historical reasons only. The reason that I am allowing cases other than `pos=0` and `pos=-1` is that I want to avoid special cases as much as possible, and other integers have a perfectly sane and intuitive interpretation in that context.

That being said, there shouldn't be a problem adding support for all kinds of things to get passed in as `pos` later. I can envision the case you describe being represented as a sequence of indices between which the new axes would be evenly divided. I also don't think this PR is the place for something like that.

eric-wieser commented:
> The reason that I am allowing cases other than pos=0 and pos=-1 is that I want to avoid special cases as much as possible, and other integers have a perfectly sane and intuitive interpretation in that context.

The other integers have sane, and intuitive meanings in both context, but unfortunately the meanings are incompatible. Does the argument represent the position to insert the **existing** axes, or the **new** axes, into the final shape?

madphysicist commented:
The position of the new axes. I see where you were getting the split version now. I think that the expectation is generally that the new axes will all be bunched together.

Erotemic commented:
@eric-wieser In general, I shy away from inplace operations because it makes debugging easier. However, I think they have their place as long as it is explicit as to when they occur. That being said, I think having an `inplace` kwarg that defaults to `False` is fine. Its off by default, but its there if you know what you are doing. In the past, I have used inplace modification of `.shape`, so I can see this option being of use. 

I would absolutely expect the new axis to be grouped together. Remember, the primary use case of this function will be when you input is typically already n dimensions, but sometimes something with fewer than n dimensions (looking at you gray-scale images) or an empty array can sneak in. In these cases there is only one new dimension or all dimensions are inserted (so ordering doesn't matter). I think we should be catering to those cases with this function and not over-design it for a use case that hasn't presented itself.

madphysicist commented:
@Erotemic Given that I have the same primary use case in mind, it may be worth considering a broadcast option instead of inserting unit dimensions. That way you'd have your RGB image directly from the grayscale.

madphysicist commented:
@Erotemic Given that I have the same primary use case in mind, it may be worth considering a broadcast option instead of inserting unit dimensions. That way you'd have your RGB image directly from the grayscale.

Erotemic commented:
It's been awhile since I've actually battled this problem, but IIRC there was an instance where if I applied an operation to a 3D image it worked, but a 2D image would not work. Perhaps I was broadcasting incorrectly, but either way I think it's immensely useful to be guaranteed that a third channel dimension exists for grayscale images, even if its only size 1. It certainly makes checks for alpha channels a lot simpler and overall reduces on the need for conditional statements.

shoyer commented on numpy/core/shape_base.py:
```diff
@@ -179,6 +171,99 @@ def atleast_3d(*arys):
+    To process multiple arrays, use a comprehension or loop around the
+    function call. See examples below.
+
+    Setting ``pos='left'`` is equivalent to how the array would be
```
needs an update

shoyer commented on numpy/core/shape_base.py:
```diff
@@ -179,6 +171,99 @@ def atleast_3d(*arys):
+        The minimum number of dimensions required.
+    pos : int, optional
+        The index of the new dimensions. All new dimensions are inserted
+        into the specified position. 0 and -1 always guaranteed to work.
```
Can you clarify what valid positions are? Something like "specified position of axes on the input array"?

shoyer commented:
A couple of comments on the API:
- Let's call the parameter for specifying the number of dimension `ndim`. `atleast_nd(array, ndim=3)` is self-explanatory in a way that `atleast_nd(array, n=3)` is not. Many people will skip the keyword argument, but if you're going to use it anyways it's better to have something descriptive. NumPy has too many functions already with non-descriptive argument names.
- I am -1 on an `inplace` argument, for numerous reasons:
     - As @eric-wieser notes, modifying `shape` in place is an antipattern, which should be discouraged as much as possible.
    - None of the other array manipulation functions in NumPy have an inplace argument.
    - It's not clear if what the return value is now. Is it `None` like most inplace functions or the original `array`?
    - There is nothing to be gained from an efficiency or code expressiveness perspective from using `atleast_nd(array, 3, inplace=True)` rather than `array = atleast_nd(array, 3)`

madphysicist commented:
@shoyer. I've squashed in all the changes you requested. `inplace` really does not make sense when you put it that way.

eric-wieser commented on numpy/core/shape_base.py:
```diff
@@ -179,6 +171,100 @@ def atleast_3d(*arys):
+        The minimum number of dimensions required.
+    pos : int, optional
+        The index to insert the new dimensions. May range from
+        ``-ary.ndim`` to ``+ary.ndim``. Non-negative indices indicate
```
Not true - the lower bound is `-ary.ndim - 1`

madphysicist replied:
Thanks for the catch. I had a feeling at the back of my head that the indices should not be the same, but was in too much of a hurry to place it. Fixed now.

madphysicist commented:
@eric-wieser The only use-case I can think of is something I did a while back where I needed to do a dot product over the last two dimensions in a 2D image that contained essentially small 2D matrices for each pixel, which would eventually be converted to color values (so a 4D array). After all the axis rolling, it turned out that the best way to broadcast to the particular RGB format I was interested in was to insert two unit axes at index 2. I honestly can't remember the exact details, and I'm sure there is a better way to do it, but I think it's better to split the existing dimensions rather than the new ones. I would also like to leave the ability to insert into an arbitrary index, not just 0 and -1.

madphysicist commented:
I forgot to remove the `inplace` parameter in the function signature. Fixed now. I believe that fixes the last remaining issue, until the next one.

madphysicist commented:
Rebased. Is everyone more or less happy with the state of this PR?

Erotemic commented:
@eric-wieser 

> Can you give a concrete example of where you'd expect this? As you note, for the examples you've given so far, it doesn't matter.

Say I have a 1D array with shape (N,), and I want to interpret it as an image so I can pass it to my image processing tools that all expect 3D arrays where the first two dimensions are spatial and the last are channels. To ensure that inputs are in correct formats I would use atleast_nd with n=3 and pos=-1. In this case I would expect the new shape to be (N, 1, 1). A similar case might arise when working with 4D tensor libraries like pytorch. I don't think its an entirely common case, but I think it would exist and in that rare case I would expect new axis to be grouped. 

I think the key idea that illustrates my expectation is what happens when there are 3 new axes? Will there be two at the front and one at the end or the other way. It seems ambiguous to allow for any other case than grouped axes. 

> Given that we've come up with no convincing use cases for intermediate values, perhaps we should just allow only pos=0 and pos=-1?

This is my preference, but I'm not attached to it. I just think its conceptually easier to understand document and use. It prevents users from shooting themselves in the foot.

madphysicist commented:
@Erotemic While I agree with you that grouping the new axes rather than the existing ones is by far the more sensible approach, I prefer to allow positions other than 0 and -1. For the work that I do, the color planes are stored separately, so it is often easier to have the channel number as the first rather than the last dimension. In that case it would be very helpful to insert the unit dimension into index -3.

madphysicist commented:
Bump?

charris commented:
@eric-wieser Want to revisit this? You seemed the most critical.

@madphysicist Was this discussed on the list at some point?

madphysicist commented:
@charris. At some length, but a very long time ago: https://mail.python.org/pipermail/numpy-discussion/2016-July/075722.html

ahaldane commented:
I only skimmed the discussions, but:

As syntactic sugar, I think it's a good addition. I was just looking at code that would use this: The `block` PRs happen to [define their own](https://github.com/numpy/numpy/pull/11971/files#diff-2270128d50ff15badd1aba4021c50a8cR449) `_atleast_nd` for clarity, and in `multiarraymodule.c` there is a C function `_prepend_ones` which essentially does this, used by the C code for `np.concatenate`, and for the `ndim` kwd of `np.array`.

Suggestion for this PR: Optimize for the most common case where `pos == 0`, by having a clause `if pos == 0: return np.array(a, ndmin=ndim, copy=False, subok=True)` at the start. That would make nice use of the `_prepend_ones` code path.

ahaldane commented on numpy/core/shape_base.py:
```diff
@@ -180,6 +172,100 @@ def atleast_3d(*arys):
+    [[[[1, 2, 3]]]]
+    """
+    ary = array(ary, copy=False, subok=True)
+    if ary.ndim:
```
why this if?

Why not move the `pos = ` line into the `if extra >0` block?

ahaldane commented on numpy/core/shape_base.py:
```diff
@@ -180,6 +172,100 @@ def atleast_3d(*arys):
+        pos = normalize_axis_index(pos, ary.ndim + 1)
+    extra = ndim - ary.ndim
+    if extra > 0:
+        ind = pos * (slice(None),) + extra * (None,) + (Ellipsis,)
```
This code reminds me it would be nice to have an `axis` argument to `put`: see issue https://github.com/numpy/numpy/issues/8765

charris commented:
This addition looks desirable. @madphysicist I'm going to close this, but feel free to rebase and resubmit. That will bring it closer to the head of the list :)
